src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1892,6 +1906,35 @@ void Session::send_unbind_message()
 }
 
 
+void Session::send_json_error_message()
+{
+    REALM_ASSERT(m_state == Active);
+    REALM_ASSERT(m_ident_message_sent);
+    REALM_ASSERT(!m_unbind_message_sent);
+
+    ClientProtocol& protocol = m_conn.get_client_protocol();
+    OutputBuffer& out = m_conn.get_output_buffer();
+    session_ident_type session_ident = get_ident();
+    std::error_code ec = m_client_error->code();
+    auto message = m_client_error->what();
+
+    logger.info("Sending: ERROR \"%1\" (error_code=%2, session_ident=%3)", message, ec.value(),
+                session_ident); // Throws
+
+    nlohmann::json error_body_json;
+    error_body_json["message"] = message;
+    protocol.make_json_error_message(out, session_ident, ec.value(), error_body_json.dump()); // Throws
+
+    auto handler = [this, ec]() {
```

Comment:
Is it possible for `this` to become invalid between initiating the write message and calling this handler? I think the way other parts of the sync client handle this is to add a flag that indicates the error was sent and then call `enlist_to_send()` and handle the unbind/disconnect in `send_message()`. It's pretty annoying because it means more and more state, but it may cover more corner cases.

Reply:
great suggestion